library/ssl_tls.c

```diff
@@ -5384,103 +5850,239 @@ int mbedtls_ssl_write_finished( mbedtls_ssl_context *ssl )
     }
 #endif
 
-#if defined(MBEDTLS_SSL_PROTO_DTLS)
-    if( ssl->conf->transport == MBEDTLS_SSL_TRANSPORT_DATAGRAM )
-        mbedtls_ssl_send_flight_completed( ssl );
-#endif
+    return( 0 );
+}
 
-    if( ( ret = mbedtls_ssl_write_record( ssl ) ) != 0 )
+static int ssl_finished_out_postprocess( mbedtls_ssl_context *ssl )
+{
+    /*
+     * In case of session resuming, invert the client and server
+     * ChangeCipherSpec messages order.
+     */
+    if( ssl->handshake->resume != 0 )
     {
-        MBEDTLS_SSL_DEBUG_RET( 1, "mbedtls_ssl_write_record", ret );
-        return( ret );
+#if defined(MBEDTLS_SSL_CLI_C)
+        if( ssl->conf->endpoint == MBEDTLS_SSL_IS_CLIENT )
+            mbedtls_ssl_handshake_set_state( ssl,
+                                   MBEDTLS_SSL_HANDSHAKE_WRAPUP );
+#endif
+#if defined(MBEDTLS_SSL_SRV_C)
+        if( ssl->conf->endpoint == MBEDTLS_SSL_IS_SERVER )
+            mbedtls_ssl_handshake_set_state( ssl,
+                                   MBEDTLS_SSL_CLIENT_CHANGE_CIPHER_SPEC );
+#endif
+    }
+    else
+    {
+#if defined(MBEDTLS_SSL_CLI_C)
+        if( ssl->conf->endpoint == MBEDTLS_SSL_IS_CLIENT )
+            mbedtls_ssl_handshake_set_state( ssl,
+                                   MBEDTLS_SSL_SERVER_CHANGE_CIPHER_SPEC );
+#endif
+#if defined(MBEDTLS_SSL_SRV_C)
+        if( ssl->conf->endpoint == MBEDTLS_SSL_IS_SERVER )
+            mbedtls_ssl_handshake_set_state( ssl,
+                                   MBEDTLS_SSL_FLUSH_BUFFERS );
+#endif
     }
 
-    MBEDTLS_SSL_DEBUG_MSG( 2, ( "<= write finished" ) );
+    return( 0 );
+}
+
+static int ssl_finished_out_write( mbedtls_ssl_context *ssl,
+                                   unsigned char* buf,
+                                   size_t buflen,
+                                   size_t *olen )
+{
+    if( buflen < 4 + ssl->handshake->state_local.finished_out.digest_len )
```

Comment:
Why 4+?

Reply:
That's again the (TLS) handshake header length, and here too I introduced a constant variable and added a comment.